### PR TITLE
Authenticate as

### PR DESCRIPTION
**Pull request by benrudolph**
@czue @NoahCarnahan 
finally got a bit of time to write this
elf: @dannyroberts 
screen shot 2015-06-02 at 11 59 41 pm

### CODE REVIEW

File: corehq/apps/hqadmin/forms.py [hunk 1]
```diff
@@ -1,7 +1,12 @@
 import re
+from crispy_forms.helper import FormHelper
+from crispy_forms import layout as crispy
+from crispy_forms.bootstrap import FormActions
```
Comment (dimagimon) [inline corehq/apps/hqadmin/forms.py]: F401 'FormActions' imported but unused

File: corehq/apps/hqadmin/tests/test_authenticate_as.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+from django.test import TestCase, SimpleTestCase
```
Comment (dimagimon) [inline corehq/apps/hqadmin/tests/test_authenticate_as.py]: F401 'SimpleTestCase' imported but unused

File: corehq/apps/hqadmin/tests/test_authenticate_as.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+from corehq.apps.users.models import WebUser, CommCareUser
+
+from ..forms import AuthenticateAsForm
+from ..views import AuthenticateAs
+
+
+class AuthenticateAsFormTest(TestCase):
+
+    def test_valid_data(self):
+        mobile_worker = CommCareUser.create('potter', 'harry', '123')
```
Comment (dimagimon) [inline corehq/apps/hqadmin/tests/test_authenticate_as.py]: F841 local variable 'mobile_worker' is assigned to but never used

File: corehq/apps/hqadmin/tests/test_authenticate_as.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        cls.user = WebUser.create(cls.domain, cls.username, cls.password, is_active=True)
+        cls.user.is_superuser = True
+        cls.user.save()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.user.delete()
+
+    def test_authenticate_as(self):
+        mobile_worker = CommCareUser.create('potter', 'harry', '123')
```
Comment (dimagimon) [inline corehq/apps/hqadmin/tests/test_authenticate_as.py]: F841 local variable 'mobile_worker' is assigned to but never used

**Comment by kaapstorm** [comment]
Great feature Ben! I was looking for this the other day.

File: corehq/apps/hqadmin/views.py [hunk 5]
```diff
@@ -225,6 +229,46 @@ def contact_email(request):
...
+        return reverse(self.urlname)
+
+    @property
+    def page_context(self):
+        return {
+            'hide_filters': True,
+            'form': self.authenticate_as_form
+        }
+
+    def post(self, request, *args, **kwargs):
```
Comment (czue) [inline corehq/apps/hqadmin/views.py]: i think the dispatch decorator covers this, but just confirming this doesn't also need one

**Comment by czue** [comment]
looks good to me based on quick pass. great that there are tests! it might be good to add a test that non-superuser's can't do this since that seems like the most critical thing to get riht.

**Comment by czue** [comment]
also the build failure looks real:

```
======================================================================
ERROR: test_authenticate_as (corehq.apps.hqadmin.tests.test_authenticate_as.AuthenticateAsIntegrationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/hqadmin/tests/test_authenticate_as.py", line 58, in test_authenticate_as
    mobile_worker = CommCareUser.create('potter', 'harry', '123')
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1405, in create
    commcare_user.save(**get_safe_write_kwargs())
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1355, in save
    super(CommCareUser, self).save(**params)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1227, in save
    raise self.Inconsistent("CouchUser with username %s already exists" % self.username)
Inconsistent: CouchUser with username harry already exists
```

**Comment by dannyroberts** [comment]
Is there anything wrong with this when it comes to auditing? Previously, superusers could easily view anyone's anything, but this lets them actually do actions as that person (like make a build leaving comments as if that person had written them is something trivial that comes to mind). Also, there are some domains that specifically prevent superuser access. (Not sure if that's still a thing, but I think it is.) With this, you could easily just log in as that other person and access their domain anyway. I'm torn because at the same time this feature seems super useful... Thoughts?

**Comment by benrudolph** [comment]
@czue right you are about testing superuser privs. will update with proper test and dimagimon fixes.

@dannyroberts the main purpose of this was so that we could access people's cloudcare and help them debug. i think you're right in that it's definitely a double edged sword. i kind of wanted to have a banner at the top that alerted the person if you were logged in as someone else. we could limit login to certain mobile worker domains, but it'd still be pretty easy to log in as a web user. we could also limit it to only work for mobile workers

File: corehq/apps/hqadmin/views.py (hunk 5 above)
Comment (benrudolph) [inline corehq/apps/hqadmin/views.py]: that's correct. updated test proves this

**Comment by benrudolph** [comment]
@kaapstorm yep, should be useful especially when dealing with cloudcare

**Comment by czue** [comment]
@dannyroberts interesting point. I think we could drop an event in auditcare to note that this happened (someone assumed someone else's identity) and then that is sufficient.

@benrudolph is there a reason we have web user access? That feels a bit weirder to me, though I can't pinpoint why

**Comment by benrudolph** [comment]
no reason, just didn't think to exclude it. i'm of the same opinion though. unless there is a need i think we should exclude for the time being. auditcare is something i'd log to in the code?

**Comment by czue** [comment]
yeah, we use it to record certain actions already. I *think* it should be pretty easy/self explanatory to setup, though admittedly no one currently at the organization has really touched it ever.

**Comment by benrudolph** [comment]
Ok, this should make it so we can only login as mobile users. Also added the view to the `AUDIT_VIEWS` list and validated that it appears in the auditcare db:
`{"id":"1439018b4a15734bb9aee76f93c04ae8","key":["b-------@dimagi.com",2015,6,8,14,54,26],"value":{"request_path":"/hq/admin/auth_as/","status_code":200}},
`

**Comment by dannyroberts** [comment]
Awesome, that seem legit to me. Is there an easy way to log who you're trying to authenticate as? If it logs URL params maybe one easy way to do it is to make sure that the username you're trying to authenticate as has to be sent as a URL param (rather than a POST param).

**Comment by dannyroberts** [comment]
(Sorry to pile stuff on, but I think if we have that, we're taking all the right steps on our side)

**Comment by benrudolph** [comment]
Good thoughts, on checking the DB I don't think the actual request body is logged in auditcare, so yea that would be a good to have. I'd almost rather have auditcare just log the request body instead of changing the form to a GET request. that seems wrong to me semantically, seem reasonable?

**Comment by dannyroberts** [comment]
It's up to you. Also, I wasn't suggesting changing the request to a GET request; I was suggesting passing just that parameter as a URL parameter. (You can use URL parameters on POST requests, even though they're often called "GET parameters" and you'd access it in django with `request.GET`, even on a POST.)

**Comment by benrudolph** [comment]
Ah that makes more sense to me. cool thanks

File: corehq/apps/hqadmin/views.py [hunk 6]
```diff
@@ -76,12 +78,14 @@
 from corehq.db import Session
 from corehq.elastic import parse_args_for_es, ES_URLS, run_query
 from dimagi.utils.couch.database import get_db, is_bigcouch
+from dimagi.utils.decorators.memoized import memoized
```
Comment (dimagimon) [inline corehq/apps/hqadmin/views.py]: F401 'memoized' imported but unused

**Comment by benrudolph** [comment]
Ok this now logs the username and domain in auditcare using the params. additionally if you navigate to `/hq/admin/auth_as/<username>/<domain>` it will fill out the form for you :smile:

File: corehq/apps/hqwebapp/models.py [hunk 7]
```diff
@@ -1389,12 +1389,15 @@ def sidebar_items(self):
         admin_operations = []
 
         if self.couch_user and self.couch_user.is_staff:
```
Comment (benrudolph) [inline corehq/apps/hqwebapp/models.py]: Anyone know what the use is of the `is_staff` permission. It seems superfluous if the user is already super_user

**Comment by benrudolph** [comment]
:white_check_mark: i think this is pretty much g2g

**Comment by benrudolph** [comment]
@dimagi/team-commcare-hq Use this to do a passwordless login as another mobile worker. If the user is `john` from domain `smithsonian`, you can either:
1) type `john` in the username box and `smithsonian` in the domain box
2) type `john@smithsonian.commcarehq.org` in the username box

**Comment by czue** [comment]
:clap: make sure nate sees this when it goes out so he can use/add to his list of support docs

**Comment by sravfeyn** [comment]
Awesome! :star2:

**Pull request merged**